File: android/src/main/java/io/revlabs/capacitor/securestorage/SecureStoragePlugin.java
package io.revlabs.capacitor.securestorage;

import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

/**
 * Capacitor plugin for secure storage on Android.
 * 
 * This plugin bridges JavaScript calls from the Capacitor webview
 * to the native Android secure storage implementation.
 */
@CapacitorPlugin(name = "SecureStorage")
public class SecureStoragePlugin extends Plugin {

    private SecureStorage implementation;

    /**
     * Called when the plugin is loaded.
     * Initialize our secure storage implementation with the Android context.
     */
    @Override
    public void load() {
        implementation = new SecureStorage(getContext());
    }

    /**
     * Store a value securely.
     * 
     * JavaScript call: SecureStorage.set({ key: 'key', value: 'value' })
     */
    @PluginMethod
    public void set(PluginCall call) {
        String key = call.getString("key");
        String value = call.getString("value");

        // Validate required parameters
        if (key == null) {
            call.reject("Key string must be provided");
            return;
        }

        if (value == null) {
            call.reject("Value string must be provided");
            return;
        }

        try {
            implementation.set(key, value);
            call.resolve(); // Return success with no data
        } catch (Exception e) {
            call.reject("Error storing value", e);
        }
    }

    /**
     * Retrieve a value from secure storage.
     * 
     * JavaScript call: SecureStorage.get({ key: 'key' })
     * Returns: { value: 'value' } or { value: null } if not found
     */
    @PluginMethod
    public void get(PluginCall call) {
        String key = call.getString("key");

        // Validate required parameters
        if (key == null) {
            call.reject("Key string must be provided");
            return;
        }

        try {
            String value = implementation.get(key);
            JSObject ret = new JSObject();
            ret.put("value", value); // value will be null if key not found
            call.resolve(ret);
        } catch (Exception e) {
            call.reject("Error retrieving value", e);
        }
    }

    /**
     * Remove a value from secure storage.
     * 
     * JavaScript call: SecureStorage.remove({ key: 'key' })
     */
    @PluginMethod
    public void remove(PluginCall call) {
        String key = call.getString("key");

        // Validate required parameters
        if (key == null) {
            call.reject("Key string must be provided");
            return;
        }

        try {
            implementation.remove(key);
            call.resolve(); // Return success with no data
        } catch (Exception e) {
            call.reject("Error removing value", e);
        }
    }

    /**
     * Clear all values from secure storage.
     * 
     * JavaScript call: SecureStorage.clear()
     */
    @PluginMethod
    public void clear(PluginCall call) {
        try {
            implementation.clear();
            call.resolve(); // Return success with no data
        } catch (Exception e) {
            call.reject("Error clearing storage", e);
        }
    }

    /**
     * Get all keys stored in secure storage.
     * 
     * JavaScript call: SecureStorage.keys()
     * Returns: { keys: ['key1', 'key2', ...] }
     */
    @PluginMethod
    public void keys(PluginCall call) {
        try {
            String[] keys = implementation.keys();
            JSObject ret = new JSObject();
            JSArray jsArray = new JSArray();
            
            // Convert the Java string array to a JavaScript array
            for (String key : keys) {
                jsArray.put(key);
            }
            
            ret.put("keys", jsArray);
            call.resolve(ret);
        } catch (Exception e) {
            call.reject("Error retrieving keys", e);
        }
    }
}

File: ios/Sources/SecureStoragePlugin/SecureStoragePlugin.swift
import Foundation
import Capacitor

/**
 * Capacitor plugin for secure storage on iOS.
 * 
 * This plugin bridges JavaScript calls from the Capacitor webview
 * to the native iOS Keychain implementation.
 * 
 * The iOS Keychain is a secure storage mechanism that protects
 * sensitive data using hardware-backed encryption.
 */
@objc(SecureStoragePlugin)
public class SecureStoragePlugin: CAPPlugin, CAPBridgedPlugin {
    // Identifier for the plugin
    public let identifier = "SecureStoragePlugin"
    // Name used in JavaScript
    public let jsName = "SecureStorage"
    // List of methods exposed to JavaScript
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "set", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "get", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "remove", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "clear", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "keys", returnType: CAPPluginReturnPromise)
    ]
    // Instance of our secure storage implementation
    private let implementation = SecureStorage()

    /**
     * Store a value securely in the Keychain
     * 
     * JavaScript call: SecureStorage.set({ key: 'key', value: 'value' })
     */
    @objc func set(_ call: CAPPluginCall) {
        // Validate required parameters
        guard let key = call.getString("key") else {
            call.reject("Key string must be provided")
            return
        }
        
        guard let value = call.getString("value") else {
            call.reject("Value string must be provided")
            return
        }
        
        // Store the value and check result
        if implementation.set(key, value: value) {
            call.resolve() // Return success with no data
        } else {
            call.reject("Failed to store value")
        }
    }

    /**
     * Retrieve a value from the Keychain
     * 
     * JavaScript call: SecureStorage.get({ key: 'key' })
     * Returns: { value: 'value' } or { value: null } if not found
     */
    @objc func get(_ call: CAPPluginCall) {
        // Validate required parameters
        guard let key = call.getString("key") else {
            call.reject("Key string must be provided")
            return
        }
        
        // Get the value from Keychain (may be nil if not found)
        let value = implementation.get(key)
        call.resolve([
            "value": value as Any // as Any handles nil correctly
        ])
    }

    /**
     * Remove a value from the Keychain
     * 
     * JavaScript call: SecureStorage.remove({ key: 'key' })
     */
    @objc func remove(_ call: CAPPluginCall) {
        // Validate required parameters
        guard let key = call.getString("key") else {
            call.reject("Key string must be provided")
            return
        }
        
        // Remove the value and check result
        if implementation.remove(key) {
            call.resolve() // Return success with no data
        } else {
            call.reject("Failed to remove value")
        }
    }

    /**
     * Clear all values from the Keychain for this plugin
     * 
     * JavaScript call: SecureStorage.clear()
     */
    @objc func clear(_ call: CAPPluginCall) {
        // Clear all values and check result
        if implementation.clear() {
            call.resolve() // Return success with no data
        } else {
            call.reject("Failed to clear storage")
        }
    }

    /**
     * Get all keys stored in the Keychain by this plugin
     * 
     * JavaScript call: SecureStorage.keys()
     * Returns: { keys: ['key1', 'key2', ...] }
     */
    @objc func keys(_ call: CAPPluginCall) {
        // Get all keys from Keychain
        let keys = implementation.keys()
        call.resolve([
            "keys": keys
        ])
    }
}
